core/src/providers/mistralai.rs

```diff
@@ -0,0 +1,680 @@
+    pub id: String,
+    pub object: String,
+    pub created: u64,
+    pub model: String,
+    pub choices: Vec<ChatChoice>,
+    pub usage: Option<Usage>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatDelta {
+    pub delta: Value,
+    pub finish_reason: Option<String>,
+    pub index: usize,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct ChatChunk {
+    pub choices: Vec<ChatDelta>,
+    pub created: u64,
+    pub id: String,
+    pub model: String,
+    pub object: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct Error {
+    #[serde(alias = "type")]
+    pub _type: Option<String>,
+    pub code: Option<String>,
+    pub message: String,
+    pub object: String,
+    pub param: Option<String>,
+}
+
+impl Error {
+    pub fn message(&self) -> String {
+        format!("MistralAIAPIError: [{:?}] {}", self._type, self.message,)
+    }
+
+    pub fn retryable(&self) -> bool {
+        match self.object.as_str() {
+            "error" => match self._type {
+                Some(_) => self.message.contains("retry"),
+                None => false,
+            },
+            _ => false,
+        }
+    }
+
+    pub fn retryable_streamed(&self) -> bool {
+        self.retryable()
+    }
+}
+
+pub struct MistralAILLM {
+    id: String,
+    api_key: Option<String>,
+}
+
+impl MistralAILLM {
+    pub fn new(id: String) -> Self {
+        MistralAILLM { id, api_key: None }
+    }
+
+    fn chat_uri(&self) -> Result<Uri> {
+        Ok(format!("https://api.mistral.ai/v1/chat/completions",).parse::<Uri>()?)
+    }
+
+    fn tokenizer(&self) -> Arc<Mutex<CoreBPE>> {
+        // TODO(flav): Pending further insights into Mistral tokenizer behavior.
+        return cl100k_base_singleton();
+    }
+
+    async fn streamed_chat_completion(
+        &self,
+        uri: Uri,
+        api_key: String,
+        model_id: Option<String>,
+        messages: &Vec<ChatMessage>,
+        temperature: f32,
+        top_p: f32,
+        max_tokens: i32,
+        event_sender: Option<UnboundedSender<Value>>,
+    ) -> Result<ChatCompletion> {
+        let url = uri.to_string();
+
+        let mut builder = match es::ClientBuilder::for_url(url.as_str()) {
+            Ok(b) => b,
+            Err(_) => return Err(anyhow!("Error creating streamed client to Mistral AI")),
+        };
+        builder = match builder.method(String::from("POST")).header(
+            "Authorization",
+            format!("Bearer {}", api_key.clone()).as_str(),
+        ) {
+            Ok(b) => b,
+            Err(_) => return Err(anyhow!("Error creating streamed client to Mistral AI")),
+        };
+        builder = match builder.header("Content-Type", "application/json") {
+            Ok(b) => b,
+            Err(_) => return Err(anyhow!("Error creating streamed client to Mistral AI")),
+        };
+
+        // TODO(flav): Handle `safe_mode`.
+
+        let mut body = json!({
+            "messages": messages,
```

Comment:
I presume this will break if we pass a user message. Let's add the prompting layer ala Anthrpic :+1:

Reply:
Addressed in `to_mistral_messages` function.